iio: adc: adrv9009: Propagate PLL-RF-LO clock

Brian-Pellegrino-ARKA opened the pull request:
## PR Description

Adds a feature to propagate the PLL-TRX-RF-LO clock rate to the clock system. The scale may be set in the device tree node with trx-lo-clock-scales = <mult div>;

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -6745,6 +6787,10 @@ static int adrv9009_probe(struct spi_device *spi)
 
+	adrv9009_clk_register(phy, "-trx_lo_clk", NULL, NULL,
+				CLK_GET_RATE_NOCACHE | CLK_IGNORE_UNUSED | CLK_RECALC_NEW_RATES,
+				TRX_LO_CLK);
```
alignment preferred on open parenthesis...

Brian-Pellegrino-ARKA replied:
The calls to adrv9009_clk_register right above this have their new liens aligned two tabs in, not aligned with the opening parenthesis. Would you prefer two tabs to keep it consistent or maybe all the calls should be updated to align with the opening parenthesis?

nunojsa replied:
The preferred coding style for this in linux is:

```
adrv9009_clk_register(phy, "-trx_lo_clk", NULL, NULL,
		      CLK_GET_RATE_NOCACHE | CLK_IGNORE_UNUSED | CLK_RECALC_NEW_RATES,
		      TRX_LO_CLK);
```

And yes, we should start the lines with tabs and then use spaces to adjust to the open parenthesis...

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -6570,6 +6608,10 @@ static int adrv9009_probe(struct spi_device *spi)
 
+	ret = of_clk_get_scale(spi->dev.of_node, "trx-lo", &phy->trx_lo_clkscale);
+	if (ret)
+		return -ret;
```
Why `-ret`. Is this also a mandatory property? Should we fail probe if that is not given in DT? That means older kernels with new DTs would now fail to probe...

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -507,6 +508,19 @@ static const char * const adrv9009_ilas_mismatch_table[] = {
+	int ret = TALACT_NO_ACTION;
+
+	ret = TALISE_setRfPllFrequency(phy->talDevice, pllName, rfPllLoFrequency_Hz);
+	if (pllName == TAL_RF_PLL && ret == TALACT_NO_ACTION) {
```
please avoid camelcase names like that... For the API functions and typer we need to stick with it but for driver specific stuff, please avoid it.

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -507,6 +508,19 @@ static const char * const adrv9009_ilas_mismatch_table[] = {
+	ret = TALISE_setRfPllFrequency(phy->talDevice, pllName, rfPllLoFrequency_Hz);
+	if (pllName == TAL_RF_PLL && ret == TALACT_NO_ACTION) {
+		phy->clk_priv[TRX_LO_CLK].rate = to_ccf_scaled((unsigned long long)rfPllLoFrequency_Hz, &phy->trx_lo_clkscale);
+		ret = clk_set_rate(phy->clk_priv[TRX_LO_CLK].hw.clk, phy->clk_priv[TRX_LO_CLK].rate);
```
long lines in here... the conditions are also a bit unrelated. How about the below?

```
...
if (ret)
    return ret;
if (pll_name != TAL_RF_PLL)
    return 0;

phy->clk_priv[TRX_LO_CLK].rate = to_ccf_scaled(rfPllLoFrequency_Hz, &phy->trx_lo_clkscale);
return clk_set_rate(...);

nunojsa replied:
Note in the above that you have a cast that i'm not really sure what's the purpose of it...

Brian-Pellegrino-ARKA replied:
I agree with this and will use your code, however I believe
`if (pll_name != TAL_RF_PLL)`
should return ret so the user still gets the result of TALISE_setRfPllFrequency. Do you agree?

nunojsa replied:
Well the first condition already check for != 0 (which in practice most of the times is < 0) so it's more readable to just `return 0` indicating all is good if `pll_name !=  TAL_RF_PLL`

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -5470,28 +5492,44 @@ static int adrv9009_clk_register(struct adrv9009_rf_phy *phy,
+		ret = devm_of_clk_add_hw_provider(&phy->spi->dev, adrv9009_pll_of_clk_get, phy);
+		if (ret) {
+			dev_err(&phy->spi->dev, "Failed to add PLL clock provider on devm_of_clk_add_hw_provider\n");
+			return ret;
```
Hmm, why is the above special? You should make it so that the already existing call of_clk_add_provider() is enough...

nunojsa reviewed (changes requested):
Hi,

Thanks a lot for contributing... Some comments from me

Brian-Pellegrino-ARKA commented:
> Hi,
> 
> Thanks a lot for contributing... Some comments from me

Thanks for the review! I have pushed an update that should address these comments. The update has been tested on hardware with an ADRV9009.

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -6774,22 +6763,22 @@ static int adrv9009_probe(struct spi_device *spi)
-				CLK_GET_RATE_NOCACHE | CLK_IGNORE_UNUSED | CLK_RECALC_NEW_RATES,
-				TRX_LO_CLK);
+						  CLK_GET_RATE_NOCACHE | CLK_IGNORE_UNUSED | CLK_RECALC_NEW_RATES,
+						  TRX_LO_CLK);
```
the above all look like unrelated changes...

Brian-Pellegrino-ARKA replied:
This lines the parameters up with the opening parenthesis. Since I lined my new adrv9009_clk_register call up with the parenthesis, I also lined up the calls to adrv9009_clk_register above my addition to keep consistency. Would you like just my new call lined up to parenthesis and not the ones above it?

nunojsa replied:
Is this aligned?

nunojsa reviewed (commented):
Please just rebase and have things in one commit which adds the feature. We don't really add additional commits with review feedback (that would pollute git history a lot)

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -5470,28 +5485,39 @@ static int adrv9009_clk_register(struct adrv9009_rf_phy *phy,
+		phy->clks[source] = clk;
+		break;
+	case TRX_LO_CLK:
+		ret = devm_clk_hw_register(&phy->spi->dev, &clk_priv->hw);
```
Also in here just use devm_clk_register()... I know we should use devm_clk_register() but we should be consistent. So either use the same API the driver is already using or have precursor patch changing devm_clk_register() -> devm_clk_hw_register().

Brian-Pellegrino-ARKA commented:
> Please just rebase and have things in one commit which adds the feature. We don't really add additional commits with review feedback (that would pollute git history a lot)

Thanks, I have squashed this to one commit and addressed the comment about devm_clk_register. Please let me know if you want anything changed regarding lining the adrv9009_clk_register calls up with the parenthesis.

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -5470,28 +5484,35 @@ static int adrv9009_clk_register(struct adrv9009_rf_phy *phy,
-	clk_priv->rate *= 1000;
-
-	clk = devm_clk_register(&phy->spi->dev, &clk_priv->hw);
-	phy->clks[source] = clk;
```
why removing the above and put it in every case? Apparently that's not needed...

Brian-Pellegrino-ARKA replied:
Apologies about the spacing, my editor uses 4 space tabs but this uses 8 space tabs. I have updated it to be level with the parenthesis using 8 space tabs. I have also moved the repeated logic out of the switch cases. Thanks!

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -5470,25 +5484,27 @@ static int adrv9009_clk_register(struct adrv9009_rf_phy *phy,
 		return -EINVAL;
 	}
 
-	clk_priv->rate *= 1000;
```
I think this is the only thing missing... It seems you can keep in this generic path instead of doing for every case.

Brian-Pellegrino-ARKA replied:
I don't believe the new TRX_LO_CLK should have it's rate multiplied by 1000 here, which was why I kept it in the other clock case blocks.

nunojsa replied:
Ahh sorry, failed to see that the new clock  does not scale...

nunojsa commented on drivers/iio/adc/adrv9009.c:
```diff
@@ -6570,6 +6586,11 @@ static int adrv9009_probe(struct spi_device *spi)
 
+	if (of_clk_get_scale(spi->dev.of_node, "trx-lo", &phy->trx_lo_clkscale)) {
+		phy->trx_lo_clkscale.mult = 0;
+		phy->trx_lo_clkscale.div = 0;
```
Failed to spot this before... If not given, then we should set both to 1. Like this, I believe we'll get a divide by 0 later on...

Brian-Pellegrino-ARKA replied:
Great catch thanks!
